src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -131,17 +133,32 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
                            << status.ToString();
           }
         }
-      }));
-  pending_lease_requests_.insert(scheduling_key);
+      });
+
+  if (status.ok()) {
+    pending_lease_requests_.insert(scheduling_key);
+  } else {
+    if (lease_client != local_lease_client_) {
+      // A remote request failed. Retry the worker lease request locally
+      // if it's still in the queue.
+      // TODO(swang): Fail after some number of retries?
+      RAY_LOG(ERROR) << "Retrying attempt to schedule task at remote node. Error: "
+                     << status.ToString();
+      RequestNewWorkerIfNeeded(scheduling_key);
+    } else {
+      RAY_LOG(FATAL) << "Lost connection with local raylet. Error: " << status.ToString();
+    }
+  }
 }
 
 void CoreWorkerDirectTaskSubmitter::PushNormalTask(const rpc::WorkerAddress &addr,
                                                    rpc::CoreWorkerClientInterface &client,
                                                    const SchedulingKey &scheduling_key,
-                                                   TaskSpecification &task_spec) {
+                                                   const TaskSpecification &task_spec) {
   auto task_id = task_spec.TaskId();
   auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
-  request->mutable_task_spec()->Swap(&task_spec.GetMutableMessage());
+  RAY_LOG(DEBUG) << "Pushing normal task " << task_spec.TaskId();
+  request->mutable_task_spec()->CopyFrom(task_spec.GetMessage());
```

Comment:
Do you see any issues with the `Swap`? Copying the task spec could be expensive.

Reply:
If you Swap here and the task fails, then the task data will be gone when the TaskManager attempts to retry the task.

Reply:
Okay I see. Could you please add a note here? Or probably can cache PushTaskRequest instead as you mentioned.